### ISSUE TITLE
Race condition between inject IP and resolving DNS

### ISSUE THREAD

**Issue by cjellick** [rancher/rancher#2341]
A user has experienced a problem where an nginx container that is part of a service fails to start because the nginx cannot resolve a rancher dns entry in its config.

I believe this is due to a race condition between nginx starting and attempting to resolve a host in its config and the container's Rancher IP getting injected. In short, it appears that nginx is starting, loading its config, attempting to resolve a host, failing, and shutting down before we can inject the Rancher IP. DNS resolution is failing because it hinges on the client IP being the Rancher IP and since the rancher IP isn't injected yet, rancher-dns doesnt return a good response.

Note: this is a hypothesis at this point. I've not fully proven that this is the problem.

**Comment by ibuildthecloud** [comment]
There's not a great fix for this until Docker 1.9 and that is basically when we will target addressing this.  So November timeframe.  1.9 code freeze already happened and 1.9 will be released before DockerCon EU.

Unfortunately the work around for this to add a sleep until DNS works in the container.  There are so more clever tricks we could do to fix this before 1.9 but that seems likes too much throw away work.

**Comment by blackjid** [comment]
That's exactly what I did as a workaround with a custom `nginx-entrypoint`

```
#!/bin/bash
while ! ping -c1 rails > /dev/null; do sleep 0.5; done
...
```

https://github.com/platanus/hound/blob/docker/config/docker/nginx-entrypoint.sh#L2

**Comment by bacheson** [comment]
+1 this was a really frustrating problem that wasn't immediately obvious

**Comment by SeerUK** [comment]
Should I be worried that I'm running Docker 1.9.0 and this is still happening for me?

**Rancher Server Info:**
Ubuntu 14.04.3 LTS
Component	Version
Rancher	v0.44.0
Cattle	v0.107.0
User Interface	v0.65.0
Rancher Compose	v0.5.1

**Rancher Agent Info:**
Ubuntu 14.04.3 LTS
Agent: rancher/agent:v0.8.2

I've set up a service stack for a master/slave Jenkins setup. I've got a master node, a service alias for it so it can be called "jenkins", and I plan to have 3 slave nodes. When I start the slave nodes, they run a script to download the `slave.jar` file from a given master node. They always fail to connect to the master node, and then reboot. If I boot up a container that will run continuously, and create a service link (as I do with the slave nodes) and then ping the master, it works. 

I'm just about to try the workaround solution above. As a note, I never had this problem until today, I just updated from Rancher v33 (!!!) and that had no problems with this.

**Comment by deniseschannon** [comment]
@SeerUK We haven't put in a fix for this issue even though Docker 1.9 is out. As @ibuildthecloud mentioned, we wouldn't start working on a fix until Docker 1.9 was out, but since the bug is still open, we haven't actually made a fix in Rancher.

**Comment by RVN-BR** [comment]
+1 watching this... currently put a sleep on my startup script... not ideal, if this doesnt get resolved in the long run i'll look at doing a ping or something instead

**Comment by deadbeef84** [comment]
This is very annoying, especially when using 3rd party docker images, which basically requires you to create custom images with sleep before startup. When will this be addressed?

**Comment by vincent99** [comment]
This should be working now in 1.0.0-rc3

**Comment by LLParse** [comment]
I've reproduced this in 1.0.0-rc2 and can confirm it is fixed in 1.0.0-rc3

[Here](https://github.com/LLParse/community-catalog/blob/issue-2341/templates/issue-2341/0/docker-compose.yml) is my compose test
